### ISSUE TITLE
prometheus_scrape source or prometheus_exporter sink does not expire values

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#5867]
I'm not sure if the issue here is with the sink or the source, but a user in discord brought up that the last successful scrape value for a metric scraped by the `prometheus_scrape` source was being returned even after the node went away.

https://discord.com/channels/742820443487993987/746070591097798688/796088208386818148

### Vector Version

```
vector 0.11.1 (v0.11.1 x86_64-unknown-linux-musl 2020-12-17)
```

### Vector Configuration File

```toml
[sources.prometheus]
type = "prometheus_scrape"
endpoints = ["http://localhost:9100/metrics"]

[sinks.out]
type = "prometheus_exporter"
inputs = ["prometheus"]
address = "0.0.0.0:9598" # required
```

### Debug Output
```
Jan 05 19:02:36.557  INFO vector::app: Log level is enabled. level="vector=info,codec=info,file_source=info,tower_limit=trace,rdkafka=info"
Jan 05 19:02:36.557  INFO vector::app: Loading configs. path=[("in.toml", None)]
Jan 05 19:02:36.559  INFO vector::topology: Running healthchecks.
Jan 05 19:02:36.559  INFO vector::topology: Starting source. name="prometheus"
Jan 05 19:02:36.559  INFO vector::topology: Starting sink. name="out"
Jan 05 19:02:36.559  INFO vector::topology::builder: Healthcheck: Passed.
Jan 05 19:02:36.559  INFO vector: Vector has started. version="0.11.1" git_version="v0.11.1" released="Thu, 17 Dec 2020 17:03:52 +0000" arch="x86_64"
Jan 05 19:02:36.559  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
Jan 05 19:03:06.568 ERROR source{component_kind="source" component_name=prometheus component_type=prometheus_scrape}: vector::internal_events::prometheus: HTTP request processing error. url=http://localhost:9100/metrics error=CallRequest { source: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })) }
```

### Expected Behavior

`curl`ing the `prometheus_exporter` endpoint after the `prometheus_scrape` source fails to pull metrics should result in those metrics not being returned anymore; otherwise they are stale.

### Actual Behavior

The `prometheus_exporter` endpoint continues to return the last scraped value.

### Reproduce

* Start `node_exporter` (https://prometheus.io/docs/guides/node-exporter/)
* Start `vector` with above config
* `curl localhost:9598/metrics` and note the metrics are there
* Stop the `node_exporter`
* Wait for vector to output a scrape error
* `curl localhost:9598/metrics` and note that the metrics are still there

**Comment by jszwedko** [comment]
Notably the user had issues with the following topology:

```
source node:  node_exporter -> prometheus_scrapper -> vector sink 
mon node: vector source -> prometheus exporter ->  prometheus scratcher
```

Based on my simple example, I'm assuming the `prometheus_exporter` is the culprit, but we may want to verify this topology works end-to-end to ensure their issue is resolved.

**Comment by jszwedko** [comment]
Also of note, `telegraf` seems like it might solve this through a configurable expiration:

https://github.com/influxdata/telegraf/tree/master/plugins/outputs/prometheus_client

```
  ## Expiration interval for each metric. 0 == no expiration
  # expiration_interval = "60s"
```

**Comment by jszwedko** [comment]
@binarylogic @jamtur01 Flagging this as a regression from 0.11 given that 0.11 doesn't include timestamps in the output.

**Comment by jamtur01** [comment]
Yeah I saw. We also need integration tests that confirm every component emits metrics too.

**Comment by jszwedko** [comment]
Oops, actually, I confused this issue for https://github.com/timberio/vector/issues/5744 :sweat_smile: I don't think this one is a regression.

**Comment by jszwedko** [comment]
You are correct that we should have tests that metrics are emitted though. Those are currently lacking.

**Comment by apollo13** [comment]
Hi @jszwedko, I can also reproduce this with the following configuration:
```
---
sources:
  host_metrics:
    type: host_metrics
    collectors:
      - filesystem
    scrape_interval_secs: 15

sinks:
  out:
    type: prometheus_exporter
    inputs:
      - host_metrics

api:
  enabled: true
```

Now on the same host execute:
```
for i in {0..100}; do docker run --rm -d --network=none ubuntu:latest sleep 20; done
```

Now the prometheus exporter will export 100 filesystems ala:
```
host_filesystem_free_bytes{collector="filesystem",device="overlay",filesystem="overlay",host="apollo13",mountpoint="/var/lib/docker/overlay2/cd82acfcf060afacb6a1dbcf37b348e64c0991f5ce6221685b032d048ea470d3/merged"} 17882755072 1633160870857
```
forever. On a busy machine one could consider this a **memory leak** (a build machine spwaning many docker containers and vector running for a while…)

Since those values are never updated once the container is gone, prometheus at some point during scraping will complain:
```
level=debug ts=2021-10-01T15:22:38.038Z caller=scrape.go:1582 component="scrape manager" scrape_pool=vector target=http://tclogs1dev.tcdev.prime-sign.com:9598/metrics msg="Out of bounds metric" series="host_filesystem_free_bytes{collector="filesystem",device="proc",filesystem="proc",host="some_host",mountpoint="/run/docker/netns/7fd82e11148c"}
```
followed by the number of samples dropped
```
level=warn ts=2021-10-01T15:32:23.048Z caller=scrape.go:1534 component="scrape manager" scrape_pool=vector target=some_host:9598/metrics msg="Error on ingesting samples that are too old or are too far into the future" num_dropped=257
```

**Comment by jszwedko** [comment]
Thanks for the example @apollo13 ! That would definitely be a case where you'd want the values to expire. I'll see if we can get this in the 0.18.0 release.
